### ISSUE TITLE
sort not working with Map field

### ISSUE THREAD

**Issue by forumjava** [Stratio/cassandra-lucene-index#17]
Hi,

Tried both with 2.1.6 and 2.1.8 on difference environment but sorting over Map field does not seem to work.

Please note that sorting with other c* type as well as filter/query over map fields works fine, just that sorting over Map fields does not do anything.

Created index as:
=====
CREATE CUSTOM INDEX IF NOT EXISTS <index> ON <table> (<lucene index col>) USING 'com.stratio.cassandra.lucene.Index' WITH OPTIONS = {'refresh_seconds' : '60', 'schema' : '{ fields : { <Map field>: {type: "string"}}}'};
=====

And trying to use it as:
=====
select * from <table> where <lucene index col> = '{sort: {fields: [{field:"<map field>.<one of the fields>"}]}}';
=====

Hope the usage is correct.

**Comment by adelapena** [comment]
Can you please show a full failing example, with the table schema, some rows insertion and one failing sort query? This way it would be easier to reproduce the issue.

Thanks

**Comment by forumjava** [comment]
Hi,

Below the details:

--create table
create table mytesttable (idcol int, testtextcol text, testmapcol map<text,text>, cassandra_lucene_index text, primary key (idcol));

--create index
CREATE CUSTOM INDEX IF NOT EXISTS lucene_test_idx ON mytesttable (cassandra_lucene_index) USING 'com.stratio.cassandra.lucene.Index' WITH OPTIONS = {'refresh_seconds' : '60', 'schema' : '{ fields : {testtextcol: {type: "string"}, testmapcol: {type: "string"}}}'};

--sample inserts
insert into mytesttable (idcol, testtextcol, testmapcol) values (1, 'row1', {'attb1': 'row1attb1Val', 'attb2': 'row1attb2Val', 'attb3': 'row1attb3Val'});
insert into mytesttable (idcol, testtextcol, testmapcol) values (2, 'someLongRow2Value', {'attb1': 'someLongattb1Val', 'attb2': 'attb2Val', 'attb3': 'attb3Val'});
insert into mytesttable (idcol, testtextcol, testmapcol) values (3, 'row3', {'attb1': 'row2attb1Val', 'attb2': 'row2attb2Val', 'attb3': 'row2attb3Val'});

--working query
select * from mytesttable where cassandra_lucene_index = '{sort: {fields: [{field:"testtextcol"}]}}';

--non working query
select * from mytesttable where cassandra_lucene_index = '{sort: {fields: [{field:"testmapcol.attb1"}]}}';

**Comment by adelapena** [comment]
Hi,

We have just fixed it at master branch. 

Thanks!

**Comment by forumjava** [comment]
Thanks, this works now!

**Comment by forumjava** [comment]
Can you please confirm, this was working fine with 2.1.8.1-SNAPSHOT but have started seeing below issue with 2.1.8.2.

=====
nsCassandraRWUser@cqlsh:mantest> select * from mytesttable where cassandra_lucene_index = '{sort: {fields: [{field:"testmapcol.attb1"}]}}';
InvalidRequest: code=2200 [Invalid query] message="Mapper 'testmapcol' is not sorted"
=====

**Comment by adelapena** [comment]
[Since 2.8.1](https://github.com/Stratio/cassandra-lucene-index/blob/branch-2.1.8/CHANGELOG.md#2181-31-july-2015) ``sorted`` option is ``false`` by default. You should create the index schema this way:
```
CREATE CUSTOM INDEX lucene_test_idx 
ON mytesttable (cassandra_lucene_index) 
USING 'com.stratio.cassandra.lucene.Index' 
WITH OPTIONS = {
   'refresh_seconds' : '60', 
   'schema' : '{ fields : {
      testtextcol: {type: "string", sorted:true}, 
      testmapcol: {type: "string", sorted:true}}}'
};
```

**Comment by forumjava** [comment]
Cool. Tried that and it works fine now.
Thanks!

**Comment by victorbadila** [comment]
Hi, I have also had trouble with this issue and I tried replicating what was written here. I am testing this on version 2.1.8 of Cassandra since it was mentioned here. The only difference is that I had to make the `testmapcol` field as a `map<text, text>` as `map` was deemed an unknown data type.

The revised queries are as follows:

```
--create table
create table mytesttable (idcol int, testtextcol text, testmapcol map<text, text>, cassandra_lucene_index text, primary key (idcol));

--create index
CREATE CUSTOM INDEX IF NOT EXISTS lucene_test_idx ON mytesttable (cassandra_lucene_index) USING 'com.stratio.cassandra.lucene.Index' WITH OPTIONS = {'refresh_seconds' : '60', 'schema' : '{ fields : {testtextcol: {type: "string", sorted: true}, testmapcol: {type: "string", sorted: true}}}'};

--sample inserts
insert into mytesttable (idcol, testtextcol, testmapcol) values (1, 'row1', {'attb1': 'row1attb1Val', 'attb2': 'row1attb2Val', 'attb3': 'row1attb3Val'});
insert into mytesttable (idcol, testtextcol, testmapcol) values (2, 'someLongRow2Value', {'attb1': 'someLongattb1Val', 'attb2': 'attb2Val', 'attb3': 'attb3Val'});
insert into mytesttable (idcol, testtextcol, testmapcol) values (3, 'row3', {'attb1': 'row2attb1Val', 'attb2': 'row2attb2Val', 'attb3': 'row2attb3Val'});
```

Next I tried to test the 2 queries mentioned by the op:

```
--working query
select * from mytesttable where cassandra_lucene_index = '{sort: {fields: [{field:"testtextcol"}]}}';

--non working query
select * from mytesttable where cassandra_lucene_index = '{sort: {fields: [{field:"testmapcol.attb1"}]}}';
```

While the queries didn't yield any errors, I got 0 results for both, while performing something like `select * from mytesttable;` yielded the expected 3 results.

Any idea about this?
Thanks

**Comment by ealonsodb** [comment]
Hi @victorbadila:

I have tested your queries againt one cassandra 2.1.8 1-node cluster and cassandra-lucene-index:2.1.8.6  and it works properly. 

Please note that when you crreates the index you set 'refresh_seconds' to 60. 

You can read in  2.1.8 [doc](https://github.com/Stratio/cassandra-lucene-index/blob/branch-2.1.8/doc/src/site/sphinx/documentation.rst#indexing) what it means but generaly it is the max time taken for writes to be searchable without forcing an index refresh.

So, you can:
-Lower the  'refresh_seconds' rate
-Wait 60 seconds to be refreshed.
-Query with refresh to true to refresh the Index Reader.

```
select * from mytesttable 
where cassandra_lucene_index = '{
    sort: {
        fields: [
            {field:"testmapcol.attb3"}
        ]
    }, 
    refresh: true
}';
```

Hope this helps

**Comment by victorbadila** [comment]
@ealonsodb seems to work, thanks!

**Comment by victorbadila** [comment]
Hi again,

Though the map value sorting mentioned before seemed to work for me on versions prior to 2.2 I had trouble combining it with pagination (pagination over top-k queries) in a way that would work. I tried multiple versions of cassandra-lucene combinations ranging from 2.1.8 to 3.0.

I read the docs for the different versions a number of times, I am not sure what I am missing. I will post a light example here, I last tried it on version 2.2.5 :

```
--table
drop table testtable;
create table testtable (idcol int, x map<text, text>, lucene_index text, primary key (idcol));

--index
CREATE CUSTOM INDEX IF NOT EXISTS lucene_idx
ON testtable (lucene_index) USING 'com.stratio.cassandra.lucene.Index' 
WITH OPTIONS = { 'refresh_seconds' : '10', 'schema' : '{ fields : {x: {type: "string", sorted: true} }}' };

--inserts
insert into testtable (idcol, x) values (1, {'a': '1', 'somethingelse': '1'});
insert into testtable (idcol, x) values (2, {'a': '2', 'somethingelse': '2'});
insert into testtable (idcol, x) values (3, {'a': '1', 'somethingelse': '3'});
insert into testtable (idcol, x) values (4, {'a': '2', 'somethingelse': '4'});

--query
select x from testtable where lucene_index = '{ sort: { fields: [ { field: "x$a" } ] }, refresh: true }';
```

I am expecting the results to be sorted by the a field of the x col, instead I get this:

> "{'a': '1', 'somethingelse': '1'}"
> "{'a': '2', 'somethingelse': '2'}"
> "{'a': '2', 'somethingelse': '3'}"
> "{'a': '1', 'somethingelse': '4'}"
> 

What I'm trying to accomplish would be to get this:

> "{'a': '1', 'somethingelse': '1'}"
> "{'a': '1', 'somethingelse': '4'}"
> "{'a': '2', 'somethingelse': '2'}"
> "{'a': '2', 'somethingelse': '3'}"
> 

and ideally paginate it. say I want a limit of 1 result per page, so in the end having 2 pages with one result each in which `a: 1` and the next 2 pages with 1 result each in which `a: 2`. 
Would be great if this would be working on version 2.2.5, but ultimately I am open to using any version that fully supports this (working combination of sorting, paging, and of course filtering)

**Comment by adelapena** [comment]
Sorting in maps is not properly working in 2.x, and it is not related to paging. We have plans to fix it in 2.x branch (probably for 2.2.6.1), but it won't be ready until the end of the next week. We want to port back the 3.x post processing method, which is the component failing here, because the 3.x approach is much more solid and easy to maintain than the used in 2.x.

Anyway, both sorting maps and paging over index-sorted results are available in [3.0.6.2](https://github.com/Stratio/cassandra-lucene-index/releases/tag/3.0.6.2), as well as several bug fixes and performance improvements.

Thanks for reporting.

**Comment by victorbadila** [comment]
it works great on 3.0.6.2. thanks
